datafusion/optimizer/src/decorrelate_where_in.rs

```diff
@@ -554,14 +636,12 @@ mod tests {
             .project(vec![col("customer.c_custkey")])?
             .build()?;
 
-        // Query will fail, but we can still transform the plan
         let expected = "Projection: customer.c_custkey [c_custkey:Int64]\
-        \n  LeftSemi Join: customer.c_custkey = __correlated_sq_1.o_custkey [c_custkey:Int64, c_name:Utf8]\
+        \n  LeftSemi Join:  Filter: customer.c_custkey = __correlated_sq_1.o_custkey AND customer.c_custkey = customer.c_custkey [c_custkey:Int64, c_name:Utf8]\
```

Comment:
This looks like somewhat of a regression to me (the filter has been pulled out of the subquery and into the join). Is that intended?

Reply:
No, I do not handle this special case, just treat `customer.c_custkey = customer.c_custkey` as join filter.

I think it is better to add this to the filter of outer table, then the plan will be like:
```
"Projection: customer.c_custkey [c_custkey:Int64]\
   LeftSemi Join: customer.c_custkey = __correlated_sq_1.o_custkey [c_custkey:Int64, c_name:Utf8]
      Filter: customer.c_custkey = customer.c_custkey [..]\
         TableScan: customer [c_custkey:Int64, c_name:Utf8]\
   SubqueryAlias: __correlated_sq_1 [o_custkey:Int64]\
      Projection: orders.o_custkey AS o_custkey [o_custkey:Int64]\
         TableScan: orders [o_orderkey:Int64, o_custkey:Int64, o_orderstatus:Utf8, o_totalprice:Float64;N]"
``` 

Does this the right way?

Reply:
After more thought it probably makes sense to do this decorrelation in the general way (no special case0 as you have done and then we will rely on subsequent predicate pushdown logic to push the filters back down if possible

Reply:
Yes, it is better to do by other rule.
But I find our `predicate pushdown logic` does not support these predicates(semi-join-on) now.
https://github.com/apache/arrow-datafusion/blob/dee0dd8be6745f6cb798ba56dca6c1b936d90fd6/datafusion/optimizer/src/push_down_filter.rs#L103-L115

Maybe we should support `predicate push down` for `semi-join-on`, or we need move these predicates to the outer filter, then `predicate push down` can work(Like following).
```rust
// for sql: select * from t1 where t1.t1_int in(select t2.t2_int from t2 where t1.t1_id > 10);
// the output of this rule:
Projection: t1.t1_id, t1.t1_name, t1.t1_int                                                                                                                                                                                                                                          
  Filter: CAST(t1.t1_id AS Int64) > Int64(10)                                                                                                                                                                                                                                          
     LeftSemi Join:  Filter: t1.t1_int = __correlated_sq_2.t2_int                                                                                                                                                                                                                          
       TableScan: t1                                                                                                                                                                                                                                                                       
       SubqueryAlias: __correlated_sq_2                                                                                                                                                                                                                                                      
         Projection: t2.t2_int AS t2_int                                                                                                                                                                                                                                                       
           TableScan: t2                                                                                                                                                                                                                                                            
```